frame/support/src/storage/types/key.rs

```diff
@@ -0,0 +1,158 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Storage key type.
+
+use crate::hash::{ReversibleStorageHasher, StorageHasher, Identity};
+use codec::{EncodeLike, FullCodec};
+use sp_std::prelude::*;
+
+/// A type used exclusively by storage maps as their key type.
+///
+/// The final key generated has the following form:
+/// ```nocompile
+/// Hasher1(encode(key1))
+///     ++ Hasher2(encode(key2))
+///     ++ ...
+///     ++ HasherN(encode(keyN))
+/// ```
+pub struct Key<Hasher, KeyType, NextKeyGenerator = ()>(
+	core::marker::PhantomData<(Hasher, KeyType, NextKeyGenerator)>,
+);
+
+/// A type to chain arbitrary number of hashers. Primarily used for
+/// migrating old hashers to new hashers in StorageNMap.
+pub struct KeyHasher<Hasher, NextHasherGenerator = ()>(
+	core::marker::PhantomData<(Hasher, NextHasherGenerator)>,
+);
+
+pub trait KeyHasherGenerator {
+	/// The current hasher generated by this type
+	type CurrentHasher: StorageHasher;
+	/// The next chained key hasher
+	type NextHasher: KeyHasherGenerator;
+}
+
+impl KeyHasherGenerator for KeyHasher<(), ()> {
+	type CurrentHasher = Identity;
+	type NextHasher = KeyHasher<(), ()>;
+}
+
+impl<H: StorageHasher> KeyHasherGenerator for KeyHasher<H, ()> {
+	type CurrentHasher = H;
+	type NextHasher = KeyHasher<(), ()>;
+}
+
+impl<H: StorageHasher, KHG: KeyHasherGenerator> KeyHasherGenerator for KeyHasher<H, KHG> {
+	type CurrentHasher = H;
+	type NextHasher = KeyHasher<KHG::CurrentHasher, KHG::NextHasher>;
+}
+
+/// A trait that contains the current key as an associated type and generates the next key.
+pub trait KeyGenerator {
+	type CurrentKey: EncodeLike<Self::CurrentKey>;
+	type NextKey;
```

Comment:
You should remove this.

And implement the trait for tuples using impl_for_tuples. Than you don't need to pass this extra parameter.

Reply:
Wow, I didn't even know of `impl-trait-for-tuples` before, that's awesome!

Reply:
@bkchr I've been going through the documentation on `impl_for_tuples` and sketching it out mentally how to implement it in my head, and one of the things that popped out to me is the way we add the trait bound to `final_key`. Consider the following:

```rust
fn final_key<KArg: EncodeLike<Self::CurrentKey>>(key: KArg) -> Vec<u8>;
```

How would I be able to implement this function for `Tuple`?

```rust
fn final_key<#( KArgN ),*>( #( keyN: KArgN ),* ) -> Vec<u8> { // Can you even do this?
    #( where KArgN: EncodeLike<Tuple::CurrentKey> )*
}
```

Reply:
This is a moot point actually, because my current implementation of `final_key` doesn't work as intended either due to `NextKey` already specifying the concrete type of what the next key should look like, making the trait bound only work on the 1st key, and subsequent keys MUST be the concrete type specified.

Reply:
> How would I be able to implement this function for Tuple?
> ```
> fn final_key<#( KArgN ),*>( keys: ( #( KArgN ),* ) ) -> Vec<u8> { // Can you even do this?
>    #( where KArgN: EncodeLike<Tuple::CurrentKey> )*
> }
> ```

The signature of this function is fixed as it is a trait. Thus we cannot have a variable number of generic.

We can do something like this, but we loose the ability to send a reference to the element of the tuple:
```rust
pub struct Key<Hasher, KeyType>(core::marker::PhantomData<(Hasher, KeyType)>);

pub trait KeyGenerator {
	type Key: codec::FullCodec;
	fn final_key(k: &Self::Key) -> Vec<u8>;
}

#[impl_trait_for_tuples::impl_for_tuples(10)]
impl KeyGenerator for Tuple {
	for_tuples!( type Key = ( #( Tuple::Key ),* ); );

	// I have no idea how to make use of EncodeLike here considering we need to deconstruct the tuple into each of its variant.
	fn final_key(k: &Self::Key) -> Vec<u8> {
		let mut res = Vec::new();
		for_tuples!( #( res.append(&mut Tuple::final_key(&k.Tuple)); )* );
		res
	}
}


impl<H: StorageHasher, K: FullCodec> KeyGenerator for Key<H, K> {
	type Key = K;

	fn final_key(k: &Self::Key) -> Vec<u8> {
		k.using_encoded(H::hash).as_ref().to_vec()
	}
}
```

Reply:
Don't we also implement EncodeLike for tuples? I remember doing this. This should also mean we don't need to deconstruct the keys in here and can just use EncodeLike on the full tuple.

Reply:
Yeah we do https://github.com/paritytech/parity-scale-codec/blob/f0341dabb01aa9ff0548558abb6dcc5c31c669a1/src/encode_like.rs#L146

Reply:
The problem with `EncodeLike` is that even though you can do
```rust
fn final_key<KArg: EncodeLike<Self::Key>>(key: KArg) -> Vec<u8>
```
there is no information on `KArg` that indicates it's actually a tuple, and so `for_tuples!( #( /* ... */ key.Tuple /* ... */ )* );` will fail.

If only `EncodeLike<K>` has an associated type that returns `K`...

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0018c455f84b29b91345d0670dc448ca

`TupleToIterator` should be sealed, so that it is only implemented for tuples!

`KArg: EncodeLike + TupleToIterator`

```
let mut iterator = key.to_iterator();
let mut final_key = Vec::new();
for_tuples! {
    let val = iterator.next().expect("KArg number should be equal to Key number");
    final_key.extend(Tuple::final_hash(val));
}
```

I think this should work :P

Reply:
@bkchr Unfortunately not, because `Encode` cannot be made into a trait object, per [the comment I made in the issue](https://github.com/paritytech/substrate/issues/8291#issuecomment-813154265).

Reply:
Than just use something different. For example the encoded vec or some trait around encode that just provides using_encoced.

Reply:
There are some possibilities.

Reply:
Almost there! Here's the issue that I'm having:

`EncodeLike<Self::Key>` is implemented on `A`, but `TupleToIterator` is implemented on `(A,)`. I'm unable to implement the latter on `A`, because it will cause a conflicting trait implementation on the tuples, and if I choose to change `Self::Key` to `(A,)`, that would cause existing key tuples to have the type of `((A,), (B,), (C,), ..., (N,))`, which is not ideal.

This is really close, I'm not sure how much further I can push the type system more to make it accept `EncodeLike` arguments, but I'll continue to try and see if I can think of anything else.

Reply:
Holy crud, this was a really tough cookie, but the latest commit solved it!